/*
*  Copyright (c) 2020, The OpenThread Authors.
*  Copyright 2023-2024 NXP
*  All rights reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions are met:
*  1. Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*  2. Redistributions in binary form must reproduce the above copyright
*     notice, this list of conditions and the following disclaimer in the
*     documentation and/or other materials provided with the distribution.
*  3. Neither the name of the copyright holder nor the
*     names of its contributors may be used to endorse or promote products
*     derived from this software without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
*  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
*  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
*  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
*  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
*  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
*  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
*  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
*  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*/

From f885d6c3ec275dfe2a934f8d30d2577a70ff1f46 Mon Sep 17 00:00:00 2001
From: Sylvain Potel <sylvain.potel@nxp.com>
Date: Thu, 5 Oct 2023 13:23:59 +0200
Subject: [PATCH] Issue: WSW-33067 [FC][15.4][ZB] Leader 9.2.12 Thread
 compliance test fails

[Description]
Only reproducible on ot-br host application. Retry mechanism on VALUE_SET command with STREAM_RAW property key is not well managed by OT RadioSpinel module.

[Root Cause]
=> host spi interface remains in BUSY state as the STREAM_RAW transaction failed. ot-br host application does not attempt a retry and continue its process. As a result the following command
will fail in OT_ERROR_BUSY which lead to an exit/failure of the host application.

[Fix]
In Radio Spinel interface if the current command failed in OT_ERROR_BUSY and if mTxRadioTid is expected then a retry mechanism shall be handled.
It is done by calling WaitResponse() to handle previous transaction (stream_raw) and call again the SendCommand with current command.

[Testing]
OT / OTBR sanity check
---
 src/lib/spinel/radio_spinel_impl.hpp | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/src/lib/spinel/radio_spinel_impl.hpp b/src/lib/spinel/radio_spinel_impl.hpp
index 4f0a7a5b2..3b3948a19 100644
--- a/src/lib/spinel/radio_spinel_impl.hpp
+++ b/src/lib/spinel/radio_spinel_impl.hpp
@@ -1684,6 +1684,19 @@ otError RadioSpinel<InterfaceType>::RequestV(uint32_t          command,
     VerifyOrExit(tid > 0, error = OT_ERROR_BUSY);
 
     error = SendCommand(command, aKey, tid, aFormat, aArgs);
+
+    if( mTxRadioTid && (error == OT_ERROR_BUSY) )
+    {
+        // Handle RETRY mechanism in case of VALUE_SET / STREAM_RAW failed transaction
+        error = WaitResponse();
+
+        if( error == OT_ERROR_NONE )
+        {
+            // Call again the current command. SPI interface shall be available now !
+            error = SendCommand(command, aKey, tid, aFormat, aArgs);
+        }
+    }
+
     SuccessOrExit(error);
 
     if (aKey == SPINEL_PROP_STREAM_RAW)
-- 
2.25.1

